### PR TITLE
fix for editing magnet link tracker list

### LINKED ISSUE #2571: MacOS: Trackers list disappears after deleting a tracker during announce (4455673bed)
The latest build has a bug that previously appeared a while back where the tracker list appears to be unresponsive when deleting a tracker during announce, and then moving out of the trackers inspector panel and into another  panel and back, the trackers list entirely disappears.

Steps to reproduce:

1. Add new torrent from magnet link
2. Begin download and switch to trackers window
3. Try to delete tracker you were never going to connect with by highlighting and hitting '-'
4. Tracker window doesn't appear to delete tracker and momentarily becomes unresponsive
5. Click into any other tab and return to trackers tab
6. Trackers tab is blank 

Alternatively? 1. Delete a tracker from a magnet link

It also seems to be related to this bug https://github.com/transmission/transmission/issues/2561. Transmission was behaving normally with regards to network activity. After? A massive spike in network activity, mid-session, regardless of all peers being paused, until Transmission is quit. 

Restarting? The massive spike begins again, immediately.

### PR DESCRIPTION
**Pull request by sweetppro**
Fixes #2571.

### CODE REVIEW

File: libtransmission/torrent.cc [hunk 1]
```diff
@@ -2079,14 +2079,24 @@ bool tr_torrent::setTrackerList(std::string_view text)
     auto const lock = this->unique_lock();
 
     auto announce_list = tr_announce_list();
-    if (!announce_list.parse(text) || !announce_list.save(this->torrentFile()))
+    if (!announce_list.parse(text) || !announce_list.save(this->hasMetadata() ? this->torrentFile() : this->magnetFile()))
```
Comment (ckerr) [inline libtransmission/torrent.cc:2082]:
```suggestion
    if (!announce_list.parse(text))
    {
        return false;
    }

    auto const has_metadata = this->hasMetadata();
    if (has_metadata && !announce_list.save(this->torrentFile()))
    {
        return false;
    }
```

File: libtransmission/torrent.cc [hunk 2]
```diff
@@ -2079,14 +2079,24 @@ bool tr_torrent::setTrackerList(std::string_view text)
...
     {
         return false;
     }
 
     this->metainfo_.announceList() = announce_list;
     this->markEdited();
-
+    
+    // magnet links
+    if (!this->hasMetadata()) {
```
Comment (ckerr) [inline libtransmission/torrent.cc:2091]:
```suggestion
    if (has_metadata) {
```

File: libtransmission/torrent.cc [hunk 3]
```diff
@@ -2079,14 +2079,24 @@ bool tr_torrent::setTrackerList(std::string_view text)
...
+        tr_error* error = nullptr;
+        if (!tr_ctorSaveMagnetContents(this, this->magnetFile(), &error))
+        {
+            this->setLocalError(
+                tr_strvJoin("Unable to save magnet file: ", error->message, " ("sv, std::to_string(error->code), ")"sv));
```
Comment (ckerr) [inline libtransmission/torrent.cc:2096]:
```suggestion
            this->setLocalError(fmt::format(
                _("Couldn't save '{path}': {error} ({error_code})"),
                fmt::arg("path", this->magnetFile()),
                fmt::arg("error", error->message),
                fmt::arg("error_code", error->code)));

```

File: libtransmission/announce-list.cc [hunk 4]
```diff
@@ -225,6 +225,11 @@ bool tr_announce_list::canAdd(tr_url_parsed_t const& announce)
 
 bool tr_announce_list::save(std::string const& torrent_file, tr_error** error) const
 {
+    //we process the magnet file in tr_torrent::setTrackerList
+    if (tr_strvEndsWith(torrent_file, ".magnet"sv)) {
+        return true;
+    }
+
```
Comment (ckerr) [inline libtransmission/announce-list.cc:232]: This PR makes me realize that `tr_announce_list` has a pre-existing problem with being too tightly coupled to a .torrent file. `tr_announce_list` shouldn't need to know what the internals of a `.torrent` file look like. That's the origin of the problem that's coming to a head here, which is that it _also_ shouldn't have to know about magnet files.

For now, I'd prefer if these five lines were removed to minimize the dependencies and that the caller checks for metainfo before calling save(). I'll try and clean up announce-list, but it's out-of-scope for this PR

```suggestion
```

File: libtransmission/torrent.cc (hunk 2 above)
Comment (sweetppro) [reply libtransmission/torrent.cc:2091]: looks good, although it should be:
```
if (!has_metadata) {
```

Comment (ckerr) [reply libtransmission/torrent.cc]: Yes, you're right!

**Review by ckerr — APPROVED** [review]

**Pull request merged**